src/BookingPriceEvaluator.php

```diff
@@ -0,0 +1,159 @@
+
+    /* @since [*next-version*] */
+    use CreateNotFoundExceptionCapableTrait;
+
+    /* @since [*next-version*] */
+    use CreateInvalidArgumentExceptionCapableTrait;
+
+    /* @since [*next-version*] */
+    use CreateOutOfRangeExceptionCapableTrait;
+
+    /* @since [*next-version*] */
+    use CreateRuntimeExceptionCapableTrait;
+
+    /* @since [*next-version*] */
+    use StringTranslatingTrait;
+
+    /**
+     * The services SELECT resource model.
+     *
+     * @since [*next-version*]
+     *
+     * @var SelectCapableInterface
+     */
+    protected $servicesSelectRm;
+
+    /**
+     * The expression builder.
+     *
+     * @since [*next-version*]
+     *
+     * @var object
+     */
+    protected $exprBuilder;
+
+    /**
+     * Constructor.
+     *
+     * @since [*next-version*]
+     *
+     * @param SelectCapableInterface $servicesSelectRm The services SELECT resource model.
+     * @param object                 $exprBuilder      The expression builder.
+     */
+    public function __construct(SelectCapableInterface $servicesSelectRm, $exprBuilder)
+    {
+        $this->servicesSelectRm = $servicesSelectRm;
+        $this->exprBuilder      = $exprBuilder;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
+    public function evaluate(ValueAwareInterface $ctx = null)
+    {
+        $booking = $ctx->getValue();
+
+        if (!($booking instanceof BookingInterface)) {
+            throw $this->_createRuntimeException(
+                $this->__('Cannot determine price - argument is not a booking instance'), null, null
+            );
+        }
+
+        try {
+            $booking = $this->_normalizeContainer($booking);
+        } catch (InvalidArgumentException $exception) {
+            throw $this->_createRuntimeException(
+                $this->__('Cannot determine price - booking is not a container'), null, $exception
+            );
+        }
+
+        /* @var $booking BookingInterface|ArrayAccess|ContainerInterface */
+
+        $duration  = $this->_normalizeInt($booking->getDuration());
+        $serviceId = $this->_containerGet($booking, 'service_id');
+
+        $b = $this->exprBuilder;
+
+        $condition = $b->eq($b->ef('service', 'id'), $b->lit($serviceId));
+        // EDD Bookings' services select RM only supports AND top-level expressions
+        $services  = $this->servicesSelectRm->select($b->and($condition));
+
+        if ($this->_countIterable($services) === 0) {
+            throw $this->_createRuntimeException(
+                $this->__('Cannot determine price - service ID in booking does not match a service'), null, null
+            );
+        }
+
+        $service = reset($services);
+        $lengths = $this->_containerGet($service, 'session_lengths');
+
+        foreach ($lengths as $_lengthInfo) {
+            $_length = $this->_normalizeInt($this->_containerGet($_lengthInfo, 'sessionLength'));
+
+            if ($duration === $_length) {
```

Comment:
How about a failsafe, in case there's no session with that exact length? Or is it an absolute requirement for them to match strictly?

Reply:
It's an absolute requirement. If a booking doesn't match a known session length, then what price to yield? Session lengths have price data. Nothing else has price data.